renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -131,8 +131,645 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
+
+  pipeDesc.GS.BytecodeLength = 0;
+  pipeDesc.GS.pShaderBytecode = NULL;
+  pipeDesc.DS.BytecodeLength = 0;
+  pipeDesc.DS.pShaderBytecode = NULL;
+  pipeDesc.HS.BytecodeLength = 0;
+  pipeDesc.HS.pShaderBytecode = NULL;
+  pipeDesc.CS.BytecodeLength = 0;
+  pipeDesc.CS.pShaderBytecode = NULL;
+
+  ZeroMemory(&pipeDesc.RasterizerState, sizeof(D3D12_RASTERIZER_DESC));
+  pipeDesc.RasterizerState.FillMode = D3D12_FILL_MODE_SOLID;
+  pipeDesc.RasterizerState.CullMode = D3D12_CULL_MODE_NONE;
+  pipeDesc.RasterizerState.FrontCounterClockwise = TRUE;
+
+  pipeDesc.SampleDesc.Count = 1;
+  pipeDesc.SampleDesc.Quality = 0;
+
+  pipeDesc.InputLayout.NumElements = 0;
+  pipeDesc.InputLayout.pInputElementDescs = NULL;
+  pipeDesc.IBStripCutValue = D3D12_INDEX_BUFFER_STRIP_CUT_VALUE_DISABLED;
+  pipeDesc.PrimitiveTopologyType = D3D12_PRIMITIVE_TOPOLOGY_TYPE_TRIANGLE;
+  pipeDesc.RTVFormats.NumRenderTargets = 1;
+  pipeDesc.RTVFormats.RTFormats[0] = retFmt;
+  pipeDesc.DSVFormat = DXGI_FORMAT_UNKNOWN;
+
+  ID3D12PipelineState *samplePso = NULL;
+  HRESULT hr = m_pDevice->CreatePipeState(pipeDesc, &samplePso);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create PSO for shader debugging HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(vsBlob);
+    SAFE_RELEASE(psBlob);
+    ShaderVariable lookupResult("tex", 0.0f, 0.0f, 0.0f, 0.0f);
+    output = lookupResult;
+    return true;
+  }
+
+  ID3D12GraphicsCommandListX *cmdList = m_pDevice->GetDebugManager()->ResetDebugList();
+  rs.pipe = GetResID(samplePso);
+  rs.ApplyState(m_pDevice, cmdList);
+
+  // Create a 1x1 texture to store the sample result
+  D3D12_RESOURCE_DESC rdesc;
+  ZeroMemory(&rdesc, sizeof(D3D12_RESOURCE_DESC));
+  rdesc.Dimension = D3D12_RESOURCE_DIMENSION_TEXTURE2D;
+  rdesc.Width = 1;
+  rdesc.Height = 1;
+  rdesc.DepthOrArraySize = 1;
+  rdesc.MipLevels = 0;
+  rdesc.Format = retFmt;
+  rdesc.Flags = D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET;
+  rdesc.SampleDesc.Count = 1;
+  rdesc.SampleDesc.Quality = 0;
+
+  D3D12_HEAP_PROPERTIES heapProps;
+  heapProps.Type = D3D12_HEAP_TYPE_DEFAULT;
+  heapProps.CPUPageProperty = D3D12_CPU_PAGE_PROPERTY_UNKNOWN;
+  heapProps.MemoryPoolPreference = D3D12_MEMORY_POOL_UNKNOWN;
+  heapProps.CreationNodeMask = 1;
+  heapProps.VisibleNodeMask = 1;
+
+  ID3D12Resource *pSampleResult = NULL;
+  D3D12_RESOURCE_STATES resourceState = D3D12_RESOURCE_STATE_RENDER_TARGET;
+  hr = m_pDevice->CreateCommittedResource(&heapProps, D3D12_HEAP_FLAG_NONE, &rdesc, resourceState,
+                                          NULL, __uuidof(ID3D12Resource), (void **)&pSampleResult);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create texture for shader debugging HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(vsBlob);
+    SAFE_RELEASE(psBlob);
+    return false;
+  }
+
+  // TODO: This work to create a staging buffer, transition  the render target, and call
+  // CopyTextureRegion work should go into a function in the debug manager, similar to GetBufferData
```

Comment:
Could you use `D3D12Replay::GetTextureData` for this?

Reply:
This is exactly what I was looking for. Thanks!